packages/xo-server/src/xapi/index.js

```diff
@@ -1121,15 +1122,19 @@ export default class Xapi extends XapiBase {
       }
       return defaultSr.$ref
     })
-
     // VDIs/SRs mapping
     const vdis = {}
+    const hostPbds = host.$PBDs
     const vbds = flatMap(vm.$snapshots, '$VBDs').concat(vm.$VBDs)
     for (const vbd of vbds) {
       const vdi = vbd.$VDI
       if (vbd.type === 'Disk') {
         vdis[vdi.$ref] =
-          mapVdisSrs && mapVdisSrs[vdi.$id] ? hostXapi.getObject(mapVdisSrs[vdi.$id]).$ref : getDefaultSrRef()
+          mapVdisSrs !== undefined && mapVdisSrs[vdi.$id] !== undefined
+            ? hostXapi.getObject(mapVdisSrs[vdi.$id]).$ref
+            : intersection(vdi.$SR.$PBDs, hostPbds).length === 0
```

Comment:
You can optimize by using `Set`:

```js
const hostPbds = new Set(host.PBDs)

const isConnected = vdi.$SR.PBDs.find(ref => hostPbds.has(ref))
```

You could go even further by caching SRs:

```js
const connectedSrs = new Map()
const isConnectedSr = sr => {
  let isConnected = connectedSrs.get(sr.$ref)
  if (isConnected === undefined) {
    isConnected = sr.PBDs.find(ref => hostPbds.has(ref))
    connectedSrs.set(sr.$ref, isConnected)
  }
  return sr
}
```